File: CRM-Client/src/api/agentRunLog.ts
import { z } from 'zod'
import type { PaginatedResponse } from '@/types/pagination'
import request from '@/utils/request'

const TurnOutcomeSchema = z.enum([
  'answered',
  'blocked_unwritten',
  'waiting_confirmation',
  'waiting_input',
  'written',
  'failed',
  'clarified',
])

const StepKindSchema = z.enum(['model', 'code', 'interaction', 'api', 'background'])
const StepToneSchema = z.enum(['done', 'blocked', 'skipped'])

const AgentRunLogStepSchema = z.object({
  kind: StepKindSchema,
  title: z.string().min(1),
  detail: z.string().min(1),
  tone: StepToneSchema,
}).strict()

const AgentRunLogTurnListItemSchema = z.object({
  turn_id: z.string().min(1),
  user_id: z.number().int().positive(),
  user_name: z.string().min(1).nullable().optional(),
  user_text: z.string().min(1),
  outcome: TurnOutcomeSchema,
  summary: z.string().min(1),
  quality_score: z.number().int().min(0).max(100).nullable().optional(),
  customer_name: z.string().min(1).nullable().optional(),
  model: z.string().min(1).nullable().optional(),
  created_time: z.string().min(1),
}).strict()

const AgentRunLogTurnDetailSchema = AgentRunLogTurnListItemSchema.extend({
  steps: z.array(AgentRunLogStepSchema).length(6),
}).strict()

const AgentRunLogTurnPageSchema = z.object({
  items: z.array(AgentRunLogTurnListItemSchema),
  total: z.number().int().nonnegative(),
  page: z.number().int().positive(),
  page_size: z.number().int().positive(),
  total_pages: z.number().int().nonnegative(),
}).strict()

export type TurnOutcome = z.infer<typeof TurnOutcomeSchema>
export type AgentRunLogStep = z.infer<typeof AgentRunLogStepSchema>
export type AgentRunLogTurnListItem = z.infer<typeof AgentRunLogTurnListItemSchema>
export type AgentRunLogTurnDetail = z.infer<typeof AgentRunLogTurnDetailSchema>

export const agentRunLogApi = {
  listTurns: async (params?: {
    page?: number
    page_size?: number
    user_id?: number
    outcome?: TurnOutcome
    q?: string
  }): Promise<PaginatedResponse<AgentRunLogTurnListItem>> => {
    return AgentRunLogTurnPageSchema.parse(
      await request.get<unknown>('/v1/agent/run-log/turns', { params }),
    )
  },

  getTurn: async (turnId: string): Promise<AgentRunLogTurnDetail> => {
    return AgentRunLogTurnDetailSchema.parse(
      await request.get<unknown>(`/v1/agent/run-log/turns/${turnId}`),
    )
  },
}

File: CRM-Server/app/schemas/agent_run_log.py
"""HTTP contracts for the Agent run-log inspector."""

from __future__ import annotations

from datetime import datetime  # noqa: TC003
from typing import Literal, TypeAlias

from pydantic import BaseModel, ConfigDict, Field

TurnOutcome: TypeAlias = Literal[
    "answered",
    "blocked_unwritten",
    "waiting_confirmation",
    "waiting_input",
    "written",
    "failed",
    "clarified",
]
StepKind: TypeAlias = Literal["model", "code", "interaction", "api", "background"]
StepTone: TypeAlias = Literal["done", "blocked", "skipped"]


class AgentRunLogStep(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: StepKind
    title: str = Field(min_length=1, max_length=200)
    detail: str = Field(min_length=1, max_length=2_000)
    tone: StepTone


class AgentRunLogTurnListItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    turn_id: str = Field(min_length=1, max_length=64)
    user_id: int = Field(gt=0)
    user_name: str | None = Field(default=None, min_length=1, max_length=100)
    user_text: str = Field(min_length=1, max_length=10_000)
    outcome: TurnOutcome
    summary: str = Field(min_length=1, max_length=500)
    quality_score: int | None = Field(default=None, ge=0, le=100)
    customer_name: str | None = Field(default=None, min_length=1, max_length=255)
    model: str | None = Field(default=None, min_length=1, max_length=256)
    created_time: datetime


class AgentRunLogTurnDetail(AgentRunLogTurnListItem):
    steps: list[AgentRunLogStep] = Field(min_length=6, max_length=6)
